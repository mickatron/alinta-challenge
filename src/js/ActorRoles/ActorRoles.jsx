import PropTypes from 'prop-types';
import flow from 'lodash/fp/flow';
import flatMap from 'lodash/fp/flatMap';
import filter from 'lodash/fp/filter';
import groupBy from 'lodash/fp/groupBy';
import pairs from 'lodash/fp/toPairs';
import sortBy from 'lodash/fp/sortBy';
import uniqBy from 'lodash/fp/uniqBy';
import React from 'react';

export default function ActorRoles(props) {
  const { name, roles } = props;

  return (
    <React.Fragment key={name}>
      <strong>{name}</strong>
      <ul>
        {roles.map(role => <li key={role.name}>{role.name}</li>)}
      </ul>
    </React.Fragment>
  );
}

ActorRoles.propTypes = {
  name: PropTypes.string.isRequired,
  roles: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
  })).isRequired,
};


export const parseMovieData = flow([
  sortBy('name'),
  flatMap(movie => movie.roles),
  uniqBy(role => role.name),
  filter(role => role.actor && role.name),
  groupBy('actor'),
  pairs,
  sortBy(item => item[0]),
]);