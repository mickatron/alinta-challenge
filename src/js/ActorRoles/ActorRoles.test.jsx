
import React from 'react';
import { shallow } from 'enzyme';

import ActorRoles, { parseMovieData } from './ActorRoles';

const mockData = [
  {
    name: 'Beverly Hills Cop',
    roles: [{ name: 'Axel Foley', actor: 'Eddie Murphy' }, { name: 'Billy Rosewood', actor: 'Judge Reinhold' }, { name: 'Mikey Tandino', actor: '' }],
  },
  {
    name: 'Another Movie',
    roles: [{ name: 'Axels Foley', actor: 'Eddie Murphy' }, { name: 'Billy Rosewood', actor: 'Judge Reinhold' }],
  },
];

describe('ActorRoles:', () => {
  let component = null;
  describe('renders', () => {
    it('a snapshot', () => {
      component = shallow((
        <ActorRoles
          name="Some Actor"
          roles={[
            { name: 'Axel Foley' },
            { name: 'Axels Foley' },
          ]}
        />
      ));
      expect(component).toMatchSnapshot();
    });
  });

  describe('{parseMovieData}:', () => {
    it('parses movie api data', () => {
      const parsedData = parseMovieData(mockData);
      expect(parsedData).toMatchSnapshot();
      expect(parsedData.length).toEqual(2);
      // first item has correct actor
      expect(parsedData[0][0]).toEqual('Eddie Murphy');
      // first item has correct number of roles
      expect(parsedData[0][1].length).toEqual(2);
    });
  });
});
