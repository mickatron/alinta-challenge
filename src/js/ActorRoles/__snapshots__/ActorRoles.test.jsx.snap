// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActorRoles: {parseMovieData}: parses movie api data 1`] = `
Array [
  Array [
    "Eddie Murphy",
    Array [
      Object {
        "actor": "Eddie Murphy",
        "name": "Axels Foley",
      },
      Object {
        "actor": "Eddie Murphy",
        "name": "Axel Foley",
      },
    ],
  ],
  Array [
    "Judge Reinhold",
    Array [
      Object {
        "actor": "Judge Reinhold",
        "name": "Billy Rosewood",
      },
    ],
  ],
]
`;

exports[`ActorRoles: renders a snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ActorRoles
    name="Some Actor"
    roles={
        Array [
            Object {
              "name": "Axel Foley",
            },
            Object {
              "name": "Axels Foley",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "Some Actor",
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <strong>
          Some Actor
</strong>,
        <ul>
          <li>
                    Axel Foley
          </li>
          <li>
                    Axels Foley
          </li>
</ul>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Some Actor",
        },
        "ref": null,
        "rendered": "Some Actor",
        "type": "strong",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <li>
              Axel Foley
</li>,
            <li>
              Axels Foley
</li>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "Axel Foley",
            "nodeType": "host",
            "props": Object {
              "children": "Axel Foley",
            },
            "ref": null,
            "rendered": "Axel Foley",
            "type": "li",
          },
          Object {
            "instance": null,
            "key": "Axels Foley",
            "nodeType": "host",
            "props": Object {
              "children": "Axels Foley",
            },
            "ref": null,
            "rendered": "Axels Foley",
            "type": "li",
          },
        ],
        "type": "ul",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "Some Actor",
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <strong>
            Some Actor
</strong>,
          <ul>
            <li>
                        Axel Foley
            </li>
            <li>
                        Axels Foley
            </li>
</ul>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Some Actor",
          },
          "ref": null,
          "rendered": "Some Actor",
          "type": "strong",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <li>
                Axel Foley
</li>,
              <li>
                Axels Foley
</li>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "Axel Foley",
              "nodeType": "host",
              "props": Object {
                "children": "Axel Foley",
              },
              "ref": null,
              "rendered": "Axel Foley",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "Axels Foley",
              "nodeType": "host",
              "props": Object {
                "children": "Axels Foley",
              },
              "ref": null,
              "rendered": "Axels Foley",
              "type": "li",
            },
          ],
          "type": "ul",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
